### ISSUE TITLE
error: pip install smac 

### ISSUE THREAD

**Issue by lzyjoe** [automl/SMAC3#807]
hello all, 
now im greenie and encountering a problem as below: 

plz help me out. thanks a lot  

ERROR: Command errored out with exit status 1:
   command: 'e:\users\LiuZY14\.conda\envs\SMAC\python.exe' 'e:\users\LiuZY14\.conda\envs\SMAC\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py' build_wheel 'e:\users\LiuZY14\AppData\Local\Temp\tmpqsomg1xx'
       cwd: E:\users\LiuZY14\AppData\Local\Temp\pip-install-25eoun63\configspace_316eab937ca04cc5ade801dfd5bbe59a
  Complete output (33 lines):
  E:\users\LiuZY14\AppData\Local\Temp\pip-build-env-jjnrfrzd\overlay\Lib\site-packages\setuptools\dist.py:723: UserWarning: Usage of dash-separated 'description-file' will not be supported in future versions. Please use the underscore name 'description_file' instead
    warnings.warn(
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.9
  creating build\lib.win-amd64-3.9\ConfigSpace
  copying ConfigSpace\exceptions.py -> build\lib.win-amd64-3.9\ConfigSpace
  copying ConfigSpace\__init__.py -> build\lib.win-amd64-3.9\ConfigSpace
  copying ConfigSpace\__version__.py -> build\lib.win-amd64-3.9\ConfigSpace
  creating build\lib.win-amd64-3.9\ConfigSpace\nx
  copying ConfigSpace\nx\exception.py -> build\lib.win-amd64-3.9\ConfigSpace\nx
  copying ConfigSpace\nx\release.py -> build\lib.win-amd64-3.9\ConfigSpace\nx
  copying ConfigSpace\nx\__init__.py -> build\lib.win-amd64-3.9\ConfigSpace\nx
  creating build\lib.win-amd64-3.9\ConfigSpace\read_and_write
  copying ConfigSpace\read_and_write\json.py -> build\lib.win-amd64-3.9\ConfigSpace\read_and_write
  copying ConfigSpace\read_and_write\pcs.py -> build\lib.win-amd64-3.9\ConfigSpace\read_and_write
  copying ConfigSpace\read_and_write\pcs_new.py -> build\lib.win-amd64-3.9\ConfigSpace\read_and_write
  copying ConfigSpace\read_and_write\__init__.py -> build\lib.win-amd64-3.9\ConfigSpace\read_and_write
  creating build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms
  copying ConfigSpace\nx\algorithms\cycles.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms
  copying ConfigSpace\nx\algorithms\dag.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms
  copying ConfigSpace\nx\algorithms\__init__.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms
  creating build\lib.win-amd64-3.9\ConfigSpace\nx\classes
  copying ConfigSpace\nx\classes\digraph.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\classes
  copying ConfigSpace\nx\classes\graph.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\classes
  copying ConfigSpace\nx\classes\__init__.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\classes
  creating build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms\components
  copying ConfigSpace\nx\algorithms\components\strongly_connected.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms\components
  copying ConfigSpace\nx\algorithms\components\__init__.py -> build\lib.win-amd64-3.9\ConfigSpace\nx\algorithms\components
  running build_ext
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  ----------------------------------------
  ERROR: Failed building wheel for ConfigSpace
  Building wheel for pynisher (setup.py) ... done
  Created wheel for pynisher: filename=pynisher-0.6.4-py3-none-any.whl size=7044 sha256=9af0aa82c1b4e9fe7c5e4715c417f82c9854c6569fcfe0e16bbf2c8361f1b1e7
  Stored in directory: e:\users\liuzy14\appdata\local\pip\cache\wheels\1d\de\5e\d4947b76b76ba27581d1e09f395eca1583a802203a41c04873
  Building wheel for pyrfr (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'e:\users\LiuZY14\.conda\envs\SMAC\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'E:\\users\\LiuZY14\\AppData\\Local\\Temp\\pip-install-25eoun63\\pyrfr_648645ce03444f5c908617e35fc48cab\\setup.py'"'"'; __file__='"'"'E:\\users\\LiuZY14\\AppData\\Local\\Temp\\pip-install-25eoun63\\pyrfr_648645ce03444f5c908617e35fc48cab\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'E:\users\LiuZY14\AppData\Local\Temp\pip-wheel-0c_pdvm4'
       cwd: E:\users\LiuZY14\AppData\Local\Temp\pip-install-25eoun63\pyrfr_648645ce03444f5c908617e35fc48cab\
  Complete output (8 lines):
  running bdist_wheel
  running build
  running build_ext
  building 'pyrfr._regression' extension
  swigging pyrfr/regression.i to pyrfr/regression_wrap.cpp
  swig.exe -python -c++ -modern -py3 -features nondynamic -I./include -o pyrfr/regression_wrap.cpp pyrfr/regression.i
  Deprecated command line option: -modern. This option is now always on.
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  ----------------------------------------
  ERROR: Failed building wheel for pyrfr
  Running setup.py clean for pyrfr
Successfully built pynisher
Failed to build ConfigSpace pyrfr
ERROR: Could not build wheels for ConfigSpace which use PEP 517 and cannot be installed directly

**Comment by renesass** [comment]
Hi, SMAC is not tested on Windows (see https://automl.github.io/SMAC3/master/pages/getting_started/installation.html).
I'd recommend you to use a subsystem or to use linux directly.

Best,
René

**Comment by lzyjoe** [comment]
> 
Bonjour René, 
Thanks for your prompt reply. 
I was figuring out if I can optimize design parameters with your package. 
Now I have some data on manufacturing sensors or measurement data as inputs.  I can understand your examples processes etc. but the output will provide us optimum design parameters right???
Thanks a lot 

BR
JOEY

**Comment by lzyjoe** [comment]
> Hi, SMAC is not tested on Windows (see https://automl.github.io/SMAC3/master/pages/getting_started/installation.html). I'd recommend you to use a subsystem or to use linux directly.
> 
> Best, René

Bonjour René,
Thanks for your prompt reply.
I was wondering if I can optimize design parameters with your package.
Now I have some data on manufacturing sensors or measurement data as inputs. I can understand your examples processes etc. but the output will provide us optimum design parameters right???
Thanks a lot

BR
JOEY

**Comment by renesass** [comment]
Yes. In the end, you get the best found configuration/hyperparameters.

**Comment by lzyjoe** [comment]
> Yes. In the end, you get the best found configuration/hyperparameters.

thanks! 
would you mind if you can paste an example with outputs???
thanks again!

**Comment by renesass** [comment]
I just updated the examples on the dev branch:
https://automl.github.io/SMAC3/development/pages/examples/python/plot_spear_mf_instances.html#sphx-glr-pages-examples-python-plot-spear-mf-instances-py

**Comment by lzyjoe** [comment]
> 



> I just updated the examples on the dev branch: https://automl.github.io/SMAC3/development/pages/examples/python/plot_spear_mf_instances.html#sphx-glr-pages-examples-python-plot-spear-mf-instances-py

HI rene,

I'm sorry to bother you again. 
I rent a linux cloud service and successfully installed smac.  However still confused by the complexity of functionalities (impressive works you have done. but difficult for me). I have my project due at the end of this week. I do really wish I could implement my work with your package.  hope you can answer me some questions as below:

1.) I run the Spear-QCP with some unknown errors(first run crashed. check up your setup). however Non-deterministic was run successfully. I supposed there's no issues on the package installation right?

2.) still working on finding design parameters optimum. I have a data with 7 features and one output variable. 
I used Xgboostregressor which is not bad for fitting features and the output variable. 
next, need using your package to reversely find the optimal features settings for a good output variable(we know certain values range as best expected value).

I corrected codes from you examples but totally lost.  would you mind if you could correct mines ??

import logging
logging.basicConfig(level=logging.INFO)
import numpy as np
from smac.facade.smac_ac_facade import SMAC4AC
from smac.scenario.scenario import Scenario
from smac.configspace import ConfigurationSpace
from ConfigSpace.hyperparameters import UniformFloatHyperparameter
from sklearn.model_selection import train_test_split
import xgboost as xgb

x = data_clean.drop(['impb'],axis=1)
y = data_clean['impb']
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=21)

def xgboost_from_cfg(cfg, seed=0):
   
    clf =  xgb.XGBRegressor(n_estimators=1500).fit(X_train,y_train)
    return xgb_model.score(X_train,y_train)

if __name__ == '__main__':
    # actually every feature has its own reasonable ranges(I don't know if i correctly defined them as following)
    cs = ConfigurationSpace()
    x1 = UniformFloatHyperparameter("feature1", 0, 3, default_value=1.5)
    x2 = UniformFloatHyperparameter("feature2", -5, 10, default_value=-4)
    x3= UniformFloatHyperparameter("feature3", 0, 3, default_value=1.5)
    x4= UniformFloatHyperparameter("feature4", 0, 3, default_value=1.1)
    x5= UniformFloatHyperparameter("feature5", 0, 3, default_value=1.2)
    x6= UniformFloatHyperparameter("feature6", 0, 3, default_value=1.7)
    x7= UniformFloatHyperparameter("feature7", 0, 3, default_value=2.5)
    cs.add_hyperparameters([x1,x2,x3,x4,x5,x6,x7])
    # Scenario object, (hope its's right)
    scenario = Scenario({"run_obj": "quality",  # we optimize quality (alternatively runtime)
                         "runcount-limit": 10,  # max. number of function evaluations
                         "cs": cs,  # configuration space
                         "deterministic": "true"
                         })
    # provide arguments for the intensifier like this
    intensifier_kwargs = {
        'n_seeds': 2,  # specify the number of seeds to evaluate for a non-deterministic target algorithm
        'initial_budget': 1,
        'eta': 3,
        'min_chall': 1  # because successive halving cannot handle min_chall > 1
    }
    smac = SMAC4AC(
        scenario=scenario,  # scenario object
        intensifier_kwargs=intensifier_kwargs,  # arguments for Successive Halving
        # change intensifier to successive halving by passing the class.
        # it must implement `AbstractRacer`.
        intensifier=SuccessiveHalving
    )
    # Start optimization
    try:
        incumbent = smac.optimize()
    finally:
        incumbent = smac.solver.incumbent
  
    print("Optimized configuration %s" % str(incumbent))

sorry for message bombing you!

Thanks a lot 

BR
JOEY

**Comment by renesass** [comment]
Hi Joey,

please use correct text formating to better read your code next time.
It seems like your TAE (xgboost_from_cfg) does not use the configuration. Also, it would be better to pass the dataset via the partial function:
```
def xgboost_from_cfg(cfg, X_train, y_train):
    ....
tae = partial(xgboost_from_cfg, X_train=X_train, y_train=y_train, ...)
```

Best,
René
